casper/src/test/scala/coop/rchain/casper/api/LastFinalizedAPITest.scala

```diff
@@ -0,0 +1,263 @@
+          val engine = new EngineWithCasper[Task](node.casperEff)
+          for {
+            b2 <- createBlock[Task](
+                   Seq(genesis.blockHash),
+                   genesis,
+                   v3,
+                   bonds,
+                   Map(v1 -> genesis.blockHash, v2 -> genesis.blockHash, v3 -> genesis.blockHash)
+                 )
+            b3 <- createBlock[Task](
+                   Seq(genesis.blockHash),
+                   genesis,
+                   v1,
+                   bonds,
+                   Map(v1 -> genesis.blockHash, v2 -> genesis.blockHash, v3 -> genesis.blockHash)
+                 )
+            b4 <- createBlock[Task](
+                   Seq(b3.blockHash),
+                   genesis,
+                   v2,
+                   bonds,
+                   Map(v1 -> b3.blockHash, v2 -> genesis.blockHash, v3 -> genesis.blockHash)
+                 )
+            b5 <- createBlock[Task](
+                   Seq(b4.blockHash),
+                   genesis,
+                   v1,
+                   bonds,
+                   Map(v1 -> b3.blockHash, v2 -> b4.blockHash, v3 -> genesis.blockHash)
+                 )
+            b6 <- createBlock[Task](
+                   Seq(b5.blockHash),
+                   genesis,
+                   v2,
+                   bonds,
+                   Map(v1 -> b5.blockHash, v2 -> b4.blockHash, v3 -> genesis.blockHash)
+                 )
+            engineCell        <- Cell.mvarCell[Task, Engine[Task]](engine)
+            dag               <- node.blockDagStorage.getRepresentation
+            lastFinalizedHash <- node.lastFinalizedBlockCalculator.run(dag, genesis.blockHash)
+            _                 = lastFinalizedHash shouldBe b4.blockHash
+            isFinalized <- BlockAPI.isFinalized[Task](
+                            IsFinalizedQuery(hash = ProtoUtil.hashString(b3))
+                          )(Sync[Task], engineCell, SafetyOracle[Task], blockStore, logEff)
+            _ = isFinalized.right.value shouldBe IsFinalizedResponse(isFinalized = true)
+          } yield ()
+      }
+    }
+  }
+
+  "isFinalized" should "return false for child of last finalized block" in effectTest {
+    HashSetCasperTestNode.standaloneEff(genesisContext).use { node =>
+      IndexedBlockDagStorage.create(node.blockDagStorage).flatMap {
+        implicit indexedBlockDagStorage =>
+          val genesis = node.genesis
+          val bonds   = ProtoUtil.bonds(genesis)
+          import node.{blockStore, casperState, cliqueOracleEffect, logEff}
+          val engine = new EngineWithCasper[Task](node.casperEff)
+          for {
+            b2 <- createBlock[Task](
+                   Seq(genesis.blockHash),
+                   genesis,
+                   v3,
+                   bonds,
+                   Map(v1 -> genesis.blockHash, v2 -> genesis.blockHash, v3 -> genesis.blockHash)
+                 )
```

Comment:
Why are you steering block creation by hand instead of just simulating real node behaviour by using 3 nodes? Those tests would be like 10x shorter with that approach.

See https://github.com/rchain/rchain/blob/dc1f0ef633618ff0671975ee88c6709e851286e6/casper/src/test/scala/coop/rchain/casper/EstimatorHelperTest.scala#L47-L77 for an example of creating a complex graph (yous seems simpler :))

Reply:
Good point. I rewrote the PR tests by using three real nodes instead. Unfortunately, many older tests in `coop.rchain.casper.api` follow the "manual" block creation pattern.